File: Awesome/Notifications/AWObserver.m
//
//  AWObserver.m
//  Awesome
//
//  Thing wrapper around the AXObserver functions. Done this way because doing
//  this from swift was really difficult to figure out, and this way "just works".
#import "AWObserver.h"

void watcher_observer_callback(AXObserverRef observer, AXUIElementRef element, CFStringRef notification, void* contextData) {
    AWObserver * ob = (__bridge AWObserver *) contextData;
    ob.callback(observer, element, notification);
}

@implementation AWObserver

- (id) init: (pid_t) pid callback:(Callback) fn  {
    if ( self = [super init] ) {
        AXObserverRef newObserver = NULL;
        // TODO: check err object, not sure what the right action should be if it failed.
        // Throw an exception?
        AXError err = AXObserverCreate(pid, watcher_observer_callback, &newObserver);
        self.callback = fn;
        self.observer = newObserver;
        
        // add the observer to the run loop, this may need to happen after we add notifications
        CFRunLoopSourceRef source = AXObserverGetRunLoopSource(newObserver);
        CFRunLoopAddSource(CFRunLoopGetCurrent(),
                           source,
                           kCFRunLoopDefaultMode);
    }
    return self;
}

- (void) addNotification:(AXUIElementRef) element notification:(CFStringRef) notifo {
    AXObserverAddNotification(self.observer, element, notifo, (__bridge void *) self);
}

-(void) removeNotification: (AXUIElementRef) element notification:(CFStringRef) notifo {
    AXObserverRemoveNotification(self.observer, element, notifo);
}

@end

File: Awesome/Notifications/AWObserver.swift
//
//  AWObserver.swift
//  Awesome
//
//  Created by Marshall Brekka on 11/23/15.
//  Copyright © 2015 Marshall Brekka. All rights reserved.
//

import Foundation

typealias Callback = (ob:AXObserverRef, el:AXUIElementRef, notification:CFStringRef) -> Void

class AWObserver {
    let callback:Callback
    let observer:AXObserver
    var selfPointer:UnsafeMutablePointer<Void>?
    
    
    init(pid:pid_t, callback:Callback) {
        var observer:AXObserver? = nil
        let err = AXObserverCreate(pid, {(observer:AXObserverRef, element:AXUIElementRef, notification:CFStringRef, contextData:UnsafeMutablePointer<Void>) in
            
            let me = Unmanaged<AWObserver>.fromOpaque(COpaquePointer(contextData)).takeUnretainedValue()
            me.callback(ob: observer, el: element, notification: notification)
        }, &observer)

        self.callback = callback
        self.observer = observer!
        self.selfPointer = UnsafeMutablePointer(Unmanaged.passUnretained(self).toOpaque())
        let source = AXObserverGetRunLoopSource(observer!)
        CFRunLoopAddSource(CFRunLoopGetCurrent(), source.takeUnretainedValue(), kCFRunLoopDefaultMode)
    }
    
    func addNotification(element:AXUIElementRef, notification:CFStringRef) {
        AXObserverAddNotification(self.observer, element, notification, selfPointer!)
    }
    
    deinit {
       let source = AXObserverGetRunLoopSource(observer)
        CFRunLoopRemoveSource(CFRunLoopGetCurrent(),
            source.takeUnretainedValue(),
            kCFRunLoopDefaultMode);
    }
}
